memberlist.go

```diff
@@ -191,50 +191,111 @@ func (m *Memberlist) Join(existing []string) (int, error) {
 		}
 
 	}
-
 	if numSuccess > 0 {
 		errs = nil
 	}
 	return numSuccess, errs
 }
 
+// ipPort holds information about a node we want to try to join.
+type ipPort struct {
+	ip   []byte
+	port uint16
+}
+
+// tcpLookupIP is a helper to initiate a TCP-based DNS lookup for the given host.
+// The built-in Go resolver will do a UDP lookup first, and will only use TCP if
+// the response has the truncate bit set, which isn't common on DNS servers like
+// Consul's. By doing the TCP lookup directly, we get the best chance for the
+// largest list of hosts to join. Since joins are relatively rare events, it's ok
+// to do this rather expensive operation.
+func (m *Memberlist) tcpLookupIP(host string, defaultPort uint16) ([]ipPort, error) {
+	cc, err := dns.ClientConfigFromFile(m.config.DNSConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(cc.Servers) > 0 {
+		server := cc.Servers[0]
+		if !hasPort(server) {
+			server = net.JoinHostPort(server, cc.Port)
+		}
+
+		c := new(dns.Client)
+		c.Net = "tcp"
+		m := new(dns.Msg)
+		m.SetQuestion(host, dns.TypeANY)
+		in, _, err := c.Exchange(m, server)
+		if err != nil {
+			return nil, err
+		}
+
+		var ips []ipPort
+		for _, r := range in.Answer {
+			switch rr := r.(type) {
+			case (*dns.A):
+				ips = append(ips, ipPort{rr.A, defaultPort})
+			case (*dns.AAAA):
+				ips = append(ips, ipPort{rr.AAAA, defaultPort})
+			}
```

Comment:
It might be worth handling `CNAME` or emitting a log message if there's a logger in scope about CNAME not being supported that response type comes back.  Intra-organization DNS shouldn't be using CNAMEs, but it's not unheard of.  The common case when talking to Consul doesn't have this problem, however.